lib/auth/apiserver.go

```diff
@@ -341,6 +349,208 @@ func (s *APIServer) deleteReverseTunnel(auth ClientI, w http.ResponseWriter, r *
+}
+
+func (s *APIServer) enableTrustedCluster(auth ClientI, tc services.TrustedCluster, version string) error {
+	vtcr := ValidateTrustedClusterRequest{
+		Token: tc.GetToken(),
+		CAs:   [][]byte{},
+	}
+
+	for _, caType := range []services.CertAuthType{services.HostCA} {
+		ca, err := auth.GetCertAuthorities(caType, false)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, sca := range ca {
+			data, err := services.GetCertAuthorityMarshaler().MarshalCertAuthority(sca, services.WithVersion(version))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			vtcr.CAs = append(vtcr.CAs, data)
+		}
+
+	}
+
+	authServer := s.APIConfig.AuthServer
+	vres, err := authServer.RequestValidateTrustedCluster(tc.GetProxyAddress(), vtcr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// the remote cluster has verified our token, take the authorities and reverse
+	// tunnels it replied with and add them to our backend
+	for _, rca := range vres.CAs {
+		ca, err := services.GetCertAuthorityMarshaler().UnmarshalCertAuthority(rca)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if ca.GetType() == services.UserCA {
+			for _, r := range tc.GetRoles() {
+				ca.AddRole(r)
+			}
+		}
+
+		if err := auth.UpsertCertAuthority(ca, backend.Forever); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	nrt := services.NewReverseTunnel(tc.GetName(), []string{tc.GetReverseTunnelAddress()})
+	err = auth.UpsertReverseTunnel(nrt, backend.Forever)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (s *APIServer) disableTrustedCluster(auth ClientI, tc services.TrustedCluster) error {
+	err := auth.DeleteCertAuthority(services.CertAuthID{Type: services.UserCA, DomainName: tc.GetName()})
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+
+	err = auth.DeleteCertAuthority(services.CertAuthID{Type: services.HostCA, DomainName: tc.GetName()})
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+
+	err = auth.DeleteReverseTunnel(tc.GetName())
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (s *APIServer) upsertTrustedCluster(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	var req *upsertTrustedClusterReq
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tc, err := services.GetTrustedClusterMarshaler().Unmarshal(req.TrustedCluster)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if tc.GetEnabled() {
+		s.enableTrustedCluster(auth, tc, version)
+	} else {
+		s.disableTrustedCluster(auth, tc)
+	}
+
+	err = auth.UpsertTrustedCluster(tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return message("ok"), nil
+}
+
+func (s *APIServer) validateTrustedCluster(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
```

Comment:
It doesn't actually register anything, this function is run on the "remote" server that checks (validates) that it generated the token.